src/vehicle/vehicle.go

```diff
@@ -1,126 +1,172 @@
 	"github.com/HyperloopUPV-H8/Backend-H8/sniffer"
 	"github.com/HyperloopUPV-H8/Backend-H8/unit_converter"
-	"github.com/HyperloopUPV-H8/Backend-H8/vehicle/internals"
 	"github.com/HyperloopUPV-H8/Backend-H8/vehicle/models"
+	"github.com/HyperloopUPV-H8/Backend-H8/vehicle/packet_parser"
 	"github.com/rs/zerolog"
 )
 
 type Vehicle struct {
-	sniffer          sniffer.Sniffer
-	parser           packet_parser.PacketParser
-	messageParser    message_parser.MessageParser
+	sniffer sniffer.Sniffer
+	pipes   map[string]*pipe.Pipe
+
 	displayConverter unit_converter.UnitConverter
 	podConverter     unit_converter.UnitConverter
-	pipes            map[string]*pipe.Pipe
 
-	packetFactory internals.UpdateFactory
+	packetParser     packet_parser.PacketParser
+	protectionParser ProtectionParser
+	bitarrayParser   BitarrayParser
 
-	updateChan  chan []byte
-	messageChan chan []byte
+	dataChan chan packet.Packet
 
 	idToBoard map[uint16]string
 
-	stats              Stats
-	statsMx            *sync.Mutex
 	onConnectionChange func(string, bool)
 
 	trace zerolog.Logger
 }
 
-func (vehicle *Vehicle) Listen(updateChan chan<- models.Update, messagesChan chan<- interface{}) {
-	vehicle.trace.Info().Msg("start listening")
-	go func() {
-		for raw := range vehicle.updateChan {
-			rawCopy := make([]byte, len(raw))
-			copy(rawCopy, raw)
+func (vehicle *Vehicle) Listen(updateChan chan<- models.PacketUpdate, protectionChan chan<- models.Protection) {
+	vehicle.trace.Debug().Msg("vehicle listening")
+	for raw := range vehicle.dataChan {
+		payloadCopy := make([]byte, len(raw.Payload))
+		copy(payloadCopy, raw.Payload)
+
+		switch id := raw.Metadata.ID; {
+		case vehicle.packetParser.Ids.Has(id):
+			update, err := vehicle.packetParser.Decode(id, raw.Payload, raw.Metadata)
 
-			id, fields := vehicle.parser.Decode(rawCopy)
-			fields = vehicle.podConverter.Revert(fields)
-			fields = vehicle.displayConverter.Convert(fields)
+			if err != nil {
+				vehicle.trace.Error().Err(err).Msg("error decoding packet")
+				continue
+			}
 
-			update := vehicle.packetFactory.NewUpdate(id, rawCopy, fields)
-			vehicle.statsMx.Lock()
-			vehicle.stats.recv++
-			vehicle.statsMx.Unlock()
+			convertedValues := vehicle.applyUnitConversion(update.Values)
+			update.Values = convertedValues
 
-			vehicle.trace.Trace().Msg("read")
 			updateChan <- update
-		}
-	}()
 
-	go func() {
-		for raw := range vehicle.messageChan {
-			msg, err := vehicle.messageParser.Parse(raw)
+		case vehicle.protectionParser.Ids.Has(id):
+			protection, err := vehicle.protectionParser.Parse(id, raw.Payload)
+
 			if err != nil {
-				vehicle.trace.Error().Stack().Err(err).Str("raw", fmt.Sprintf("%#v", string(raw))).Msg("parse message")
+				vehicle.trace.Error().Err(err).Msg("error decoding protection")
 				continue
 			}
-			messagesChan <- msg
+			protectionChan <- protection
+		default:
+			fmt.Println("UNEXPECTED VALUE")
 		}
-	}()
 
+	}
 }
 
 func (vehicle *Vehicle) SendOrder(order models.Order) error {
```

Comment:
missing bitarray encoding (even if FW isn't using it put the code and we can comment it if we don't want it)